services/cloudapi/cloudapiserver/server.go

```diff
@@ -298,33 +347,102 @@ func (ss *Service) Stop(ctx context.Context, datacenterID, serverID, serverType
 		if err != nil {
 			return err
 		}
-		return utils.WaitForResourceToBeReady(ctx, ss.D, ss.checkExpectedVmStateFn(ctx, datacenterID, constant.VMStateStop))
+		if err = utils.WaitForResourceToBeReady(ctx, ss.D, ss.checkExpectedVMStateFn(datacenterID, constant.VMStateStop)); err != nil {
+			return err
+		}
+		if errState := cloudapi.WaitForStateChange(ctx, ss.Meta, ss.D, apiResponse, schema.TimeoutUpdate); errState != nil {
+			return fmt.Errorf("an error occurred while waiting for server state change on VM SHUTOFF dcId: %s, server_id: %s, Response: (%w)", datacenterID, serverID, errState)
+		}
+		log.Printf("[DEBUG] %s server powered off: serverId: %s \n", serverType, serverID)
+		return nil
 
 	case constant.CubeType:
 		apiResponse, err := ss.Client.ServersApi.DatacentersServersSuspendPost(ctx, datacenterID, serverID).Execute()
 		apiResponse.LogInfo()
 		if err != nil {
 			return err
 		}
-		return utils.WaitForResourceToBeReady(ctx, ss.D, ss.checkExpectedVmStateFn(ctx, datacenterID, constant.CubeVMStateStop))
+		if err = utils.WaitForResourceToBeReady(ctx, ss.D, ss.checkExpectedVMStateFn(datacenterID, constant.CubeVMStateStop)); err != nil {
+			return err
+		}
+		if errState := cloudapi.WaitForStateChange(ctx, ss.Meta, ss.D, apiResponse, schema.TimeoutUpdate); errState != nil {
+			return fmt.Errorf("an error occurred while waiting for server state change on VM SUSPEND dcId: %s, server_id: %s, Response: (%w)", datacenterID, serverID, errState)
+		}
+		log.Printf("[DEBUG] %s server suspended: serverId: %s \n", serverType, serverID)
+		return nil
 	}
 
 	return fmt.Errorf("cannot stop unknown server type: %s", serverType)
 
 }
 
-// checkExpectedVmStateFn wraps over the ResourceReadyFunc to allow passing expectedState
-// TODO: change ResourceReadyFunc sig to support passing an expectedState param
-func (ss *Service) checkExpectedVmStateFn(ctx context.Context, dcId, expectedState string) utils.ResourceReadyFunc {
+// Reboot resets the VM,
+func (ss *Service) Reboot(ctx context.Context, datacenterID, serverID string) error {
+
+	apiResponse, err := ss.Client.ServersApi.DatacentersServersRebootPost(ctx, datacenterID, serverID).Execute()
+	apiResponse.LogInfo()
+	if err != nil {
+		return err
+	}
+	if err = utils.WaitForResourceToBeReady(ctx, ss.D, ss.checkExpectedVMStateFn(datacenterID, constant.VMStateStart)); err != nil {
+		return err
+	}
+	if errState := cloudapi.WaitForStateChange(ctx, ss.Meta, ss.D, apiResponse, schema.TimeoutUpdate); errState != nil {
+		return fmt.Errorf("an error occurred while waiting for server state change on reboot dcId: %s, server_id: %s, Response: (%w)", datacenterID, serverID, errState)
+	}
+	log.Printf("[DEBUG] server reboot finished: serverId: %s \n", serverID)
```

Comment:
there is mixed usage of this across the project, but I agree it should be removed since it's unnecessary ([01e5d73](https://github.com/ionos-cloud/terraform-provider-ionoscloud/pull/521/commits/01e5d73e9d1987a641e3e72b81443aecf4d40a05))